File: Automation/001-AutomationFramework/C#/AutomationFramework/Common/SharedTestSteps/EmailServices/EmailServiceTestSteps.cs
using CommonCore;
using CommonCore.Tests;
using SharedTestSteps.Contracts;
using Utilities;

namespace SharedTestSteps.EmailServices
{
    public class EmailServiceTestSteps : IEmailService
    {
        private IEmailServiceOperator _emailService = null;

        public void Initialize()
        {
            switch (TestInMemoryParameters.Instance.EmailServiceName)
            {
                case "Mailinator":
                    _emailService = new MailinatorTestSteps();
                    break;
                case "Sharklasers":
                    _emailService = new SharkLasersTestSteps();
                    break;

                default:
                    throw new InvalidOperationException("Email service not initialized");
            }
        }

        public void GetVerificationCodeFromEmail()
        {
            _emailService.GetVerificationCodeFromEmail();
        }

        public void GenerateEmail()
        {
            _emailService.GenerateEmail();
        }

        public void GetEmailAddress()
        {
            _emailService.GetEmailAddress();
        }

        public virtual void CheckAvailability(List<EmailServiceParameters> services)
        {
            var service = services.Find(x => WebsiteManager.IsWebsiteAvailable(x.Url));

            if (service == null)
            {
                throw new InvalidOperationException("None of the email providers is currently available");
            }

            TestInMemoryParameters.Instance.EmailServiceUrl = service.Url;
            TestInMemoryParameters.Instance.EmailServiceName = service.Name;
        }

        public void VerifyEmailNotificationContent()
        {
            _emailService.VerifyEmailNotificationContent();
        }

        public void DeleteAllEmailsFromInbox()
        {
            _emailService.DeleteAllEmailsFromInbox();
        }
    }
}

File: Automation/Woolsocks/TestAutomationFramework/src/test/java/testSuites/shared/steps/emailServices/EmailServiceTestSteps.java
package testSuites.shared.steps.emailServices;

import common.core.EmailServiceParameters;
import common.core.TestInMemoryParameters;
import common.utilities.LoggingManager;
import common.utilities.WebsiteManager;
import testSuites.shared.steps.contracts.EmailService;

import java.util.List;
import java.util.stream.Collectors;
public class EmailServiceTestSteps extends EmailService {
    public void initialize() throws IllegalStateException {
        switch (TestInMemoryParameters.getInstance().emailServiceName) {
            case "Mailinator":
                emailService = new MailinatorTestSteps();
                LoggingManager.Info("Initialized Mailinator");
                break;
            default:
                LoggingManager.Error("Email service not initialized");
        }
    }
    public void saveGeneratedEmail() {
        emailService.saveGeneratedEmail();
    }

    public void getEmailAddress() {
        emailService.getEmailAddress();
    }

    public void checkAvailability(List<EmailServiceParameters> services) {
        var mailServices = services.stream()
                .filter(x -> WebsiteManager.isWebsiteAvailable(x.getUrl())).collect(Collectors.toList());

        if (mailServices == null) {
            LoggingManager.Error("None of the email providers is currently available");
        }

        var firstAvailableEmailService = services.stream().findFirst();
        TestInMemoryParameters.getInstance().emailServiceUrl = firstAvailableEmailService.get().url;
        TestInMemoryParameters.getInstance().emailServiceName = firstAvailableEmailService.get().name;

        LoggingManager.Info("Chosen email service provider name: " + firstAvailableEmailService.get().name);
        LoggingManager.Info("Chosen email service provider url: " + firstAvailableEmailService.get().url);
    }

    public void verifyEmailNotificationContent() {
        emailService.verifyEmailNotificationContent();
    }
}
